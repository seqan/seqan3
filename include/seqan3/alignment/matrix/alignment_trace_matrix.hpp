// ============================================================================
//                 SeqAn - The Library for Sequence Analysis
// ============================================================================
//
// Copyright (c) 2006-2018, Knut Reinert & Freie Universitaet Berlin
// Copyright (c) 2016-2018, Knut Reinert & MPI Molekulare Genetik
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
//     * Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above copyright
//       notice, this list of conditions and the following disclaimer in the
//       documentation and/or other materials provided with the distribution.
//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
//       its contributors may be used to endorse or promote products derived
//       from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
// DAMAGE.
//
// ============================================================================

/*!\file
 * \author Marcel Ehrhardt <marcel.ehrhardt AT fu-berlin.de>
 * \brief Contains the declaration of seqan3::alignment_trace_matrix.
 */

#pragma once

#include <seqan3/alignment/matrix/alignment_matrix_concept.hpp>
#include <seqan3/alignment/matrix/alignment_score_matrix.hpp>
#include <seqan3/core/add_enum_bitwise_operators.hpp>

namespace seqan3
{

/*!\brief The possible directions a trace can have. The values can be combined.
 * \ingroup alignment_matrix
 * \sa seqan3::add_enum_bitwise_operators<trace_matrix_directions> enables combining enum values.
 * \sa seqan3::alignment_trace_matrix implementations use this enum as matrix entry type.
 */
enum struct trace_matrix_directions : uint8_t
{
    //!\brief No trace
    none      = 0,
    //!\brief Trace comes from the diagonal entry.
    diagonal  = 1,
    //!\brief Trace comes from the above entry.
    up        = 2,
    //!\brief Trace comes from the left entry.
    left      = 4
};

//!\brief Enable bitwise operators for enum seqan3::trace_matrix_directions.
//!\ingroup alignment_matrix
template <>
constexpr bool add_enum_bitwise_operators<trace_matrix_directions> = true;

//!\brief The declaration of alignment_trace_matrix. Each definition of this
//!       declaration must satisfy seqan3::alignment_matrix_concept.
//!\ingroup alignment_matrix
template <typename ...>
struct alignment_trace_matrix;

/*!\brief A trace matrix represented in a one-dimensional std::vector
 * \ingroup alignment_matrix
 * \implements seqan3::alignment_matrix_concept
 * \tparam sequence_t \copydoc seqan3::alignment_matrix_concept::sequence_type
 *
 *
 * This data structure stores the matrix in a flat way using the
 * std::vector<#entry_type> data structure where each row is stored
 * continuously.
 *
 * ## Example
 *
 * \snippet test/snippet/alignment/matrix/alignment_trace_matrix_vector.cpp code
 *
 * ### Output
 * \include test/snippet/alignment/matrix/alignment_trace_matrix_vector.out
 */
template <typename sequence_t>
struct alignment_trace_matrix<std::vector<trace_matrix_directions>, sequence_t>
{
    //!\brief The type of the trace.
    using entry_type = trace_matrix_directions;

    //!\copydoc seqan3::alignment_matrix_concept::sequence_type
    using sequence_type = sequence_t;

    /*!\name Constructors, destructor and assignment
     * The copy-constructor, move-constructor, copy-assignment, move-assignment,
     * and destructor are implicitly defined.
     * \{
     */
    /*!\brief Construct the trace matrix out of the \a traces, the \a database,
     *        and the \a query.
     * \param     traces   The trace values as flat std::vector<#entry_type>.
     * \param[in] database The #database sequence.
     * \param[in] query    The #query sequence.
     *
     * \attention Make sure that the #database and the #query outlives the matrix.
     */
    alignment_trace_matrix
    (
        std::vector<entry_type> traces,
        sequence_type const & database,
        sequence_type const & query
    )
        : _traces{std::move(traces)}, _database{database}, _query{query}
    {}
    //!\}

    //!\copydoc seqan3::alignment_matrix_concept::database
    inline sequence_type const & database() const noexcept
    {
        return _database;
    }

    //!\copydoc seqan3::alignment_matrix_concept::query
    inline sequence_type const & query() const noexcept
    {
        return _query;
    }

    //!\copydoc seqan3::alignment_matrix_concept::rows
    inline std::size_t rows() const noexcept
    {
        return _query.size()+1;
    }

    //!\copydoc seqan3::alignment_matrix_concept::cols
    inline std::size_t cols() const noexcept
    {
        return _database.size()+1;
    }

    //!\copydoc seqan3::alignment_matrix_concept::at
    inline entry_type at(unsigned row, unsigned col) const noexcept
    {
        return _traces[row * cols() + col];
    }

private:
    //!\brief The matrix as a one-dimensional vector of traces
    //!       (each row is continuously stored).
    std::vector<entry_type> _traces;
    //!\copydoc seqan3::alignment_matrix_concept::database
    sequence_type const & _database;
    //!\copydoc seqan3::alignment_matrix_concept::query
    sequence_type const & _query;
};

/*!\brief A trace matrix that uses an underlying seqan3::alignment_score_matrix
 * \ingroup alignment_matrix
 * \implements seqan3::alignment_matrix_concept
 * \tparam ...score_matrix_params_t The template parameters of seqan3::alignment_score_matrix
 *
 *
 * This data structure uses directly the score matrix to infer the trace matrix
 * and works for any seqan3::alignment_score_matrix.
 *
 * \todo TODO: Is currently only able to handle the edit distance.
 *
 * ## Example
 *
 * \snippet test/snippet/alignment/matrix/alignment_trace_matrix.cpp code
 *
 * ### Output
 * \include test/snippet/alignment/matrix/alignment_trace_matrix.out
 */
template <typename ...score_matrix_params_t>
//!\cond
    requires alignment_matrix_concept<alignment_score_matrix<score_matrix_params_t...>> &&
             std::is_integral_v<typename alignment_score_matrix<score_matrix_params_t...>::entry_type>
//!\endcond
struct alignment_trace_matrix<alignment_score_matrix<score_matrix_params_t...>>
    : public alignment_score_matrix<score_matrix_params_t...>
{
    //!\brief The type of the score matrix.
    using score_matrix_type = alignment_score_matrix<score_matrix_params_t...>;

    //!\brief The type of an entry in the score matrix.
    using score_type = typename score_matrix_type::entry_type;

    //!\copydoc seqan3::alignment_matrix_concept::entry_type
    using entry_type = trace_matrix_directions;

    //!\copydoc seqan3::alignment_matrix_concept::sequence_type
    using sequence_type = typename score_matrix_type::sequence_type;

    /*!\name Constructors, destructor and assignment
     * The copy-constructor, move-constructor, copy-assignment, move-assignment,
     * and destructor are implicitly defined.
     * \{
     */
    //!\brief Construct the trace matrix by using a score_matrix.
    //!\param score_matrix The score matrix.
    alignment_trace_matrix(score_matrix_type score_matrix)
        : score_matrix_type(std::move(score_matrix))
    {}

    //!\brief Construct the trace matrix directly via the score matrix constructor.
    //!\tparam ...args_t The type of the arguments needed to construct the #score_matrix_type
    //!\param  args      The constructor arguments of the base class #score_matrix_type
    template <typename ...args_t>
    alignment_trace_matrix(args_t && ...args)
        : score_matrix_type(std::forward<args_t>(args)...)
    {}
    //!\}

    //!\copydoc seqan3::alignment_matrix_concept::rows
    using score_matrix_type::rows;
    //!\copydoc seqan3::alignment_matrix_concept::cols
    using score_matrix_type::cols;
    //!\copydoc seqan3::alignment_matrix_concept::database
    using score_matrix_type::database;
    //!\copydoc seqan3::alignment_matrix_concept::query
    using score_matrix_type::query;

    //!\brief The trace directions of the matrix at position (\a row, \a col).
    inline entry_type at(unsigned row, unsigned col) const noexcept
    {
        entry_type direction{};

        if(is_trace_diagonal(row, col))
            direction |= entry_type::diagonal;

        if(is_trace_up(row, col))
            direction |= entry_type::up;

        if(is_trace_left(row, col))
            direction |= entry_type::left;

        return direction;
    }

    //!\brief Access to the score_matrix.
    inline score_matrix_type const & score_matrix() const noexcept
    {
        return *this;
    }

private:

    //!\brief Does the trace come from the above entry?
    inline bool is_trace_up(unsigned row, unsigned col) const noexcept
    {
        // TODO: use the alignment_config to calculate the score
        score_type gap = 1;

        score_type curr = score_matrix().at(row, col);
        score_type up = row == 0 ? col : score_matrix().at(row-1, col);
        return curr == up + gap;
    }

    //!\brief Does the trace come from the left entry?
    inline bool is_trace_left(unsigned row, unsigned col) const noexcept
    {
        // TODO: use the alignment_config to calculate the score
        score_type gap = 1;

        score_type curr = score_matrix().at(row, col);
        score_type left = col == 0 ? row : score_matrix().at(row, col-1);
        return curr == left + gap;
    }

    //!\brief Does the trace come from the diagonal entry?
    inline bool is_trace_diagonal(unsigned row, unsigned col) const noexcept
    {
        // TODO: use the alignment_config to calculate the score
        score_type match = 0;
        score_type mismatch = 1;

        score_type curr = score_matrix().at(row, col);
        if (col == 0 || row == 0)
            return false;

        score_type diag = score_matrix().at(row-1, col-1);
        bool is_match = query()[row-1] == database()[col-1];

        return (is_match && curr == diag + match) ||
              (!is_match && curr == diag + mismatch);
    }
};

/*!\name Type deduction guides
 * \relates seqan3::alignment_trace_matrix
 * \{
 */
template <typename sequence_t>
alignment_trace_matrix(std::vector<trace_matrix_directions>, sequence_t const &, sequence_t const &)
    -> alignment_trace_matrix<std::vector<trace_matrix_directions>, sequence_t const &>;

template <typename ...args_t>
alignment_trace_matrix(args_t&& ...args)
    -> alignment_trace_matrix<decltype(alignment_score_matrix{std::forward<args_t>(args)...})>;

template <typename alignment_t, typename ...options_t>
alignment_trace_matrix(alignment_score_matrix<alignment_t, options_t...>)
    -> alignment_trace_matrix<alignment_score_matrix<alignment_t, options_t...>>;
//!\}

} // namespace seqan3
