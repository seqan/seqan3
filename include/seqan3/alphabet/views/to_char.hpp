// SPDX-FileCopyrightText: 2006-2024 Knut Reinert & Freie Universität Berlin
// SPDX-FileCopyrightText: 2016-2024 Knut Reinert & MPI für molekulare Genetik
// SPDX-License-Identifier: BSD-3-Clause

/*!\file
 * \brief Provides seqan3::views::to_char.
 * \author Hannes Hauswedell <hannes.hauswedell AT fu-berlin.de>
 */

#pragma once

#include <ranges>

#include <seqan3/alphabet/concept.hpp>
#include <seqan3/utility/views/deep.hpp>

namespace seqan3::views
{
/*!\brief               A view that calls seqan3::to_char() on each element in the input range.
 * \tparam urng_t       The type of the range being processed. See below for requirements. [template parameter is
 *                      omitted in pipe notation]
 * \param[in] urange    The range being processed. [parameter is omitted in pipe notation]
 * \returns             A range of converted elements. See below for the properties of the returned range.
 * \ingroup alphabet_views
 *
 * \details
 *
 * \header_file{seqan3/alphabet/views/to_char.hpp}
 *
 * ### View properties
 *
 * This view is a **deep view** Given a range-of-range as input (as opposed to just a range), it will apply
 * the transformation on the innermost range (instead of the outermost range).
 *
 * | Concepts and traits              | `urng_t` (underlying range type)      | `rrng_t` (returned range type)                                |
 * |----------------------------------|:-------------------------------------:|:-------------------------------------------------------------:|
 * | std::ranges::input_range         | *required*                            | *preserved*                                                   |
 * | std::ranges::forward_range       |                                       | *preserved*                                                   |
 * | std::ranges::bidirectional_range |                                       | *preserved*                                                   |
 * | std::ranges::random_access_range |                                       | *preserved*                                                   |
 * | std::ranges::contiguous_range    |                                       | *lost*                                                        |
 * |                                  |                                       |                                                               |
 * | std::ranges::viewable_range      | *required*                            | *guaranteed*                                                  |
 * | std::ranges::view                |                                       | *guaranteed*                                                  |
 * | std::ranges::sized_range         |                                       | *preserved*                                                   |
 * | std::ranges::common_range        |                                       | *preserved*                                                   |
 * | std::ranges::output_range        |                                       | *lost*                                                        |
 * | seqan3::const_iterable_range     |                                       | *preserved*                                                   |
 * |                                  |                                       |                                                               |
 * | std::ranges::range_reference_t   | seqan3::alphabet                      | seqan3::alphabet_char_t<std::ranges::range_value_t<urng_t>>   |
 *
 * See the \link views views submodule documentation \endlink for detailed descriptions of the view properties.
 *
 * ### Example
 * \include test/snippet/alphabet/views/range_view_to_char.cpp
 * \hideinitializer
 *
 * \stableapi{Since version 3.1.}
 */
inline auto const to_char = deep{std::views::transform(
    [](auto const in) noexcept
    {
        static_assert(alphabet<std::remove_cvref_t<decltype(in)>>,
                      "The value type of seqan3::views::to_char must model the seqan3::alphabet.");
        return seqan3::to_char(in);
    })};

} // namespace seqan3::views
