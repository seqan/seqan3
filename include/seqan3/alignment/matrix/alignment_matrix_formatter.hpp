// ============================================================================
//                 SeqAn - The Library for Sequence Analysis
// ============================================================================
//
// Copyright (c) 2006-2018, Knut Reinert & Freie Universitaet Berlin
// Copyright (c) 2016-2018, Knut Reinert & MPI Molekulare Genetik
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
//     * Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above copyright
//       notice, this list of conditions and the following disclaimer in the
//       documentation and/or other materials provided with the distribution.
//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
//       its contributors may be used to endorse or promote products derived
//       from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
// DAMAGE.
//
// ============================================================================

/*!\file
 * \author Marcel Ehrhardt <marcel.ehrhardt AT fu-berlin.de>
 * \brief Contains seqan3::alignment_matrix_formatter and seqan3::alignment_matrix_format.
 */

#pragma once

#include <iomanip>
#include <iostream>
#include <optional>
#include <sstream>

#include <seqan3/alignment/matrix/alignment_matrix_concept.hpp>
#include <seqan3/alignment/matrix/alignment_trace_matrix.hpp>
#include <seqan3/core/metafunction/template_inspection.hpp>

namespace seqan3
{

/*!\brief Format used by seqan3::alignment_matrix_formatter
 * \ingroup alignment_matrix
 *
 *
 * With seqan3::alignment_matrix_format you can style:
 *  * the epsilon symbol (alignment_matrix_format::epsilon)
 *  * the column symbol that separates each cell in a row (alignment_matrix_format::col_sep)
 *  * the row symbol that divides each row (alignment_matrix_format::row_sep)
 *  * the column symbol that comes after the row symbol (alignment_matrix_format::row_col_sep)
 *  * the traces symbols of a traceback matrix (alignment_matrix_format::trace_dir)
 *
 * ## Example
 * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.cpp alignment_matrix_format
 *
 * ### Output
 * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.out alignment_matrix_format::out
 */
struct alignment_matrix_format
{
    //!\brief epsilon symbol (a single symbol)
    char const * epsilon{};
    //!\brief column separator symbol (a single symbol)
    char const * col_sep{};
    //!\brief row separator symbol (a single symbol)
    char const * row_sep{};
    //!\brief row separator symbol (a single symbol)
    char const * row_col_sep{};

    /*!\brief eight symbols for each combination of directions a trace can have (each entry can have multiple symbols)
     *
     * * 1st bit: **D** = diagonal
     * * 2nd bit: **U** = up
     * * 3th bit: **L** = left
     *
     * | i     | 1st bit | 2nd bit | 3rd bit | trace_dir[i] |
     * | :---: |---------|---------|---------| :----------: |
     * | **0** | 0       | 0       | 0       | **No dir**   |
     * | **1** | 1       | 0       | 0       | **D**        |
     * | **2** | 0       | 1       | 0       | **U**        |
     * | **3** | 1       | 1       | 0       | **DU**       |
     * | **4** | 0       | 0       | 1       | **L**        |
     * | **5** | 1       | 0       | 1       | **DL**       |
     * | **6** | 0       | 1       | 1       | **UL**       |
     * | **7** | 1       | 1       | 1       | **DUL**      |
     *
     */
    char const * trace_dir[8]{};

    //!\brief how many bytes needs the epsilon symbol (necessary when using utf-8)
    uint8_t epsilon_bytes{1};

    /*!\brief how many bytes needs every trace_dir symbol (necessary when using utf-8)
     * \attention A restriction is that every symbol needs the same amount of
     * bytes to represent one char.
     */
    uint8_t trace_dir_bytes{1};

    /*!\brief The CSV format that makes it easy to export the matrix
     * \sa https://en.wikipedia.org/wiki/Comma-separated_values
     *
     *
     * ### Score output
     * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.out alignment_matrix_format::csv#score
     *
     * ### Trace output
     * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.out alignment_matrix_format::csv#trace
     */
    static const alignment_matrix_format csv;

    /*!\brief A format that uses only ascii symbols.
     * \sa https://en.wikipedia.org/wiki/Ascii
     *
     *
     * ### Score output
     * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.out alignment_matrix_format::ascii#score
     *
     * ### Trace output
     * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.out alignment_matrix_format::ascii#trace
     */
    static const alignment_matrix_format ascii;

    /*!\brief A format that uses unicode block symbols.
     * \sa https://en.wikipedia.org/wiki/Unicode
     *
     *
     * ### Score output
     * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.out alignment_matrix_format::unicode_block#score
     *
     * ### Trace output
     * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.out alignment_matrix_format::unicode_block#trace
     */
    static const alignment_matrix_format unicode_block;

    /*!\brief A format that uses unicode braille symbols.
     * \sa https://en.wikipedia.org/wiki/Unicode
     *
     *
     * ### Score output
     * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.out alignment_matrix_format::unicode_braille#score
     *
     * ### Trace output
     * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.out alignment_matrix_format::unicode_braille#trace
     */
    static const alignment_matrix_format unicode_braille;

   /*!\brief A format that uses unicode arrow symbols.
    * \sa https://en.wikipedia.org/wiki/Unicode
    *
    *
    * ### Score output
    * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.out alignment_matrix_format::unicode_arrows#score
    *
    * ### Trace output
    * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.out alignment_matrix_format::unicode_arrows#trace
    */
    static const alignment_matrix_format unicode_arrows;
};

constexpr alignment_matrix_format alignment_matrix_format::csv
{
    " ", ";", "", "",
    {"N","D","U","DU","L","DL","UL","DUL"}
};

constexpr alignment_matrix_format alignment_matrix_format::ascii
{
    " ", "|", "-", "/",
    {" ","D","U","DU","L","DL","UL","DUL"} // 1byte per char
};

constexpr alignment_matrix_format alignment_matrix_format::unicode_block
{
    u8"ε", u8"║", u8"═", u8"╬",
    {u8"█",u8"▘",u8"▝",u8"▀",u8"▖",u8"▌",u8"▞",u8"▛"}, // 3bytes per char
    2, 3
};

constexpr alignment_matrix_format alignment_matrix_format::unicode_braille
{
    u8"ε", u8"║", u8"═", u8"╬",
    {u8"⠀",u8"⠁",u8"⠈",u8"⠉",u8"⠄",u8"⠅",u8"⠌",u8"⠍"}, // 3bytes per char
    2, 3
};

constexpr alignment_matrix_format alignment_matrix_format::unicode_arrows
{
    u8"ε", u8"║", u8"═", u8"╬",
    {u8"↺",u8"↖",u8"↑",u8"↖↑",u8"←",u8"↖←",u8"↑←",u8"↖↑←"}, // 3bytes per char
    2, 3
};

/*!\brief Formats and prints trace and score matrices that satisfy the
 *        seqan3::alignment_matrix_concept.
 * \ingroup alignment_matrix
 * \tparam alignment_matrix_t The matrix class which satisfies the seqan3::alignment_matrix_concept
 *
 *
 * ## Example
 *
 * \snippet test/snippet/alignment/matrix/alignment_score_matrix.cpp code
 *
 * ### Output
 * \include test/snippet/alignment/matrix/alignment_score_matrix.out
 */
template <alignment_matrix_concept alignment_matrix_t>
struct alignment_matrix_formatter
{
    //!\brief The type of the #matrix
    using alignment_matrix_type = alignment_matrix_t;

    //!\brief The matrix to format.
    alignment_matrix_type const & matrix;

    //!\brief The actual format used by the formatter.
    alignment_matrix_format symbols{alignment_matrix_format::unicode_arrows};

    //!\brief Whether #alignment_matrix_type is a traceback matrix.
    //!\hideinitializer
    static constexpr bool is_traceback_matrix = []
    {
        using trace_matrix_t = detail::transfer_template_args_onto_t<alignment_matrix_type, alignment_trace_matrix>;
        return std::is_same_v<alignment_matrix_type, trace_matrix_t>;
    }();

    //!\brief Determines the largest width of all entries in the #matrix,
    //!       e.g. `-152` has width 4.
    std::size_t auto_width() const noexcept
    {
        std::size_t col_width = 1;
        for(unsigned row = 0; row < matrix.rows(); ++row)
            for(unsigned col = 0; col < matrix.cols(); ++col)
                col_width = std::max(col_width, char_length(char_at(row, col)));
        return col_width;
    }

    //!\brief Print the formatted #matrix to std::cout.
    //!\param[in]  column_width    width of each cell, std::nullopt defaults to auto_width()
    void format(std::optional<std::size_t> column_width = std::nullopt) const noexcept
    {
        format(std::cout, column_width);
    }

    /*!\brief Print the formatted #matrix to the \a cout stream.
     * \tparam         char_t          char type of std::basic_ostream
     * \tparam         traits_t        traits type of std::basic_ostream
     * \param[in,out]  cout            print formatted #matrix into this ostream
     * \param[in]      column_width    width of each cell, std::nullopt defaults to auto_width()
     */
    template <typename char_t, typename traits_t>
    void format(std::basic_ostream<char_t, traits_t> & cout, std::optional<std::size_t> column_width) const noexcept
    {
        std::size_t _column_width = column_width.has_value() ? column_width.value() : auto_width();

        auto cell = [&](auto && symbol, std::optional<std::size_t> symbol_bytes = std::nullopt)
        {
            // deal with unicode chars that mess up std::setw
            std::size_t bytes = symbol_bytes.value_or(
                is_traceback_matrix ? symbols.trace_dir_bytes : 1u);
            std::size_t length_bytes = char_length_bytes(symbol);
            std::size_t length = length_bytes / bytes;
            std::size_t offset = length_bytes - length;

            cout << std::left
                      << std::setw(_column_width + offset)
                      << symbol
                      << symbols.col_sep;
        };

        auto first_cell = [&](auto && symbol)
        {
            cout << symbol << symbols.col_sep;
        };

        // |_|d|a|t|a|b|a|s|e|
        auto first_row = [&]
        {
            first_cell(" ");
            cell(symbols.epsilon, symbols.epsilon_bytes);

            for(unsigned col = 0; col < matrix.cols()-1; ++col)
                cell(matrix.database()[col], 1);
            cout << "\n";
        };

        // |-|-|-|-|-|-|-|-|-|
        auto divider = [&]
        {
            cout << " " << symbols.row_col_sep;
            for(unsigned col = 0; col < matrix.cols(); ++col)
            {
                for(unsigned i = 0; i < _column_width; ++i)
                    cout << symbols.row_sep;
                cout << symbols.row_col_sep;
            }
            cout << "\n";
        };

        first_row();
        for(unsigned row = 0; row < matrix.rows(); ++row)
        {
            if (symbols.row_sep[0] != '\0')
                divider();

            // one query letter + one row of scores / traces
            if (row == 0)
                first_cell(symbols.epsilon);
            else
                first_cell(matrix.query()[row-1]);
            for(unsigned col = 0; col < matrix.cols(); ++col)
                cell(char_at(row, col));
            cout << "\n";
        }
    }

private:

    //!\brief Same as #matrix\.at(\a row, \a col), but converts the value to
    //!       a trace symbol (alignment_matrix_format::trace_dir) if the
    //!       #matrix is a traceback matrix.
    auto char_at(unsigned row, unsigned col) const noexcept
    {
        if constexpr(is_traceback_matrix)
        {
            trace_matrix_directions direction = matrix.at(row, col);
            return symbols.trace_dir[(unsigned)(direction) % 8u];
        } else
        {
            return matrix.at(row, col);
        }
    }

    //!\brief The length of the \a str (traceback symbols are unicode aware)
    std::size_t char_length(auto && str) const noexcept
    {
        if constexpr(is_traceback_matrix)
            return char_length_bytes(str) / symbols.trace_dir_bytes;
        return char_length_bytes(str);
    }

    //!\brief The number of bytes the \a str uses
    std::size_t char_length_bytes(auto && str) const noexcept
    {
        std::stringstream stream;
        stream << str;
        return stream.str().length();
    }
};

/*!\name Type deduction guides
 * \relates seqan3::alignment_matrix_formatter
 * \{
 */
template<typename alignment_matrix_t>
alignment_matrix_formatter(alignment_matrix_t const &) -> alignment_matrix_formatter<alignment_matrix_t>;

template<typename alignment_matrix_t>
alignment_matrix_formatter(alignment_matrix_t const &, alignment_matrix_format) -> alignment_matrix_formatter<alignment_matrix_t>;
//!\}
} // namespace seqan3
