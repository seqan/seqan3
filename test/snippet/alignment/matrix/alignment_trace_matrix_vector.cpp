#include <iostream>

#include <seqan3/alphabet/nucleotide/dna4.hpp>
#include <seqan3/alignment/matrix/alignment_matrix_formatter.hpp>
#include <seqan3/alignment/matrix/alignment_trace_matrix.hpp>
#include <seqan3/range/view/to_char.hpp>

int main()
{

//! [code]
using namespace seqan3;
using namespace seqan3::literal;

std::vector<dna4> database = "AACCGGTT"_dna4;
std::vector<dna4> query = "ACGT"_dna4;

auto N = trace_matrix_directions::none;
auto D = trace_matrix_directions::diagonal;
auto U = trace_matrix_directions::up;
auto L = trace_matrix_directions::left;

alignment_trace_matrix trace_matrix
{
    std::vector
    {
        N,L,L  ,L  ,L  ,L  ,L  ,L,L  ,
        U,D,D|L,L  ,L  ,L  ,L  ,L,L  ,
        U,U,D  ,D  ,D|L,L  ,L  ,L,L  ,
        U,U,D|U,D|U,D  ,D  ,D|L,L,L  ,
        U,U,D|U,D|U,D|U,D|U,D  ,D,D|L
    },
    database,
    query
};

std::cout << "database:\t" << (trace_matrix.database() | view::to_char) << std::endl;
std::cout << "query:\t\t" << (trace_matrix.query() | view::to_char) << std::endl;
std::cout << std::endl;

std::cout << "trace_matrix: " << trace_matrix.cols() << " columns and "
          << trace_matrix.rows() << " rows" << std::endl;

// Print out the matrix.
for(unsigned row = 0u; row < trace_matrix.rows(); ++row)
{
    for(unsigned col = 0u; col < trace_matrix.cols(); ++col)
    {
        trace_matrix_directions dir = trace_matrix.at(row, col);
        if (dir == N)
            std::cout << "N";
        if ((dir & D) == D)
            std::cout << "D";
        if ((dir & U) == U)
            std::cout << "U";
        if ((dir & L) == L)
            std::cout << "L";
        std::cout << ", ";
    }
    std::cout << std::endl;
}
std::cout << std::endl;

// Prints out the matrix in a convenient way
alignment_matrix_formatter{trace_matrix}.format();
//! [code]

    return 0;
}
