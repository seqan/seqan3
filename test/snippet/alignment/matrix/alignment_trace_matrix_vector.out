database:	[A,A,C,C,G,G,T,T]
query:		[A,C,G,T]

trace_matrix: 9 columns and 5 rows
N, L, L, L, L, L, L, L, L, 
U, D, DL, L, L, L, L, L, L, 
U, U, D, D, DL, L, L, L, L, 
U, U, DU, DU, D, D, DL, L, L, 
U, U, DU, DU, DU, DU, D, D, DL, 

 ║ε ║A ║A ║C ║C ║G ║G ║T ║T ║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
ε║↺ ║← ║← ║← ║← ║← ║← ║← ║← ║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
A║↑ ║↖ ║↖←║← ║← ║← ║← ║← ║← ║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
C║↑ ║↑ ║↖ ║↖ ║↖←║← ║← ║← ║← ║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
G║↑ ║↑ ║↖↑║↖↑║↖ ║↖ ║↖←║← ║← ║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
T║↑ ║↑ ║↖↑║↖↑║↖↑║↖↑║↖ ║↖ ║↖←║
